src/bin/cargo/cli.rs

```diff
@@ -9,7 +9,19 @@ use super::commands;
 use command_prelude::*;
 
 pub fn main(config: &mut Config) -> CliResult {
-    let args = cli().get_matches_safe()?;
+    let args = match cli().get_matches_safe() {
+        Ok(args) => args,
+        Err(e) => {
+            if e.kind == clap::ErrorKind::UnrecognizedSubcommand {
+                // An unrecognized subcommand might be an external subcommand.
+                let cmd = &e.info.as_ref().unwrap()[0].to_owned();
+                return super::execute_external_subcommand(config, cmd, &[cmd, "--help"])
```

Comment:
This assumes that subcommands support `--help` argument. I think this is a reasonable assumption, but let's add 

> Cargo assumes that `cargo-${command} --help` prints a help message

line to docs here: https://github.com/rust-lang/cargo/blob/master/src/doc/src/reference/external-tools.md#custom-subcommands

Reply:
Added a small section about this assumption, and cleaned up the docs around it a bit.